src/main/java/com/microsoft/graph/authentication/TokenCredentialAuthProvider.java

```diff
@@ -0,0 +1,87 @@
+import okhttp3.Request;
+import com.microsoft.graph.exceptions.ErrorConstants.*;
+
+import java.util.List;
+
+public class TokenCredentialAuthProvider implements ICoreAuthenticationProvider , IAuthenticationProvider {
+
+    //TokenCredential expected from user
+    private TokenCredential tokenCredential;
+    //Context options which can be optionally set by the user
+    private TokenRequestContext context;
+    //Access token to be retrieved
+    private AccessToken accessToken;
+
+    /**
+     * Creates an Authentication provider using a passed in TokenCredential
+     *
+     * @param tokenCredential Credential object inheriting the TokenCredential interface used to instantiate the Auth Provider
+     * @throws AuthenticationException exception occurs if the TokenCredential parameter is null
+     */
+    public  TokenCredentialAuthProvider(TokenCredential tokenCredential) throws AuthenticationException {
+        if(tokenCredential == null) {
+            throw new AuthenticationException(new Error(Codes.InvalidArgument,
+                    String.format(Messages.NullParameter, "TokenCredential"))
+                    ,new IllegalArgumentException());
+        }
+
+        this.tokenCredential = tokenCredential;
+        this.context = new TokenRequestContext();
+    }
+
+    /**
+     * Creates an Authentication provider using a TokenCredential and list of scopes
+     *
+     * @param tokenCredential Credential object inheriting the TokenCredential interface used to instantiate the Auth Provider
+     * @param scopes Specified desired scopes of the Auth Provider
+     * @throws AuthenticationException exception occurs if the TokenCredential parameter is null
+     */
+    public TokenCredentialAuthProvider(TokenCredential tokenCredential, List<String> scopes) throws AuthenticationException {
+        this(tokenCredential);
+        this.context.setScopes(scopes);
+    }
+
+    /**
+     * Authenticates the request
+     *
+     * @param request the request to authenticate
+     */
+    @Override
+    public void authenticateRequest(IHttpRequest request) {
+        request.addHeader(AuthConstants.AUTHORIZATION_HEADER, AuthConstants.BEARER + getAccessToken());
+    }
+
+    /**
+     * Authenticates the request
+     *
+     * @param request the request to authenticate
+     * @return Request with Authorization header added to it
+     */
+    @Override
+    public Request authenticateRequest(Request request) {
+        return request.newBuilder()
+                    .addHeader(AuthConstants.AUTHORIZATION_HEADER, AuthConstants.BEARER + getAccessToken())
+                    .build();
+    }
+
+    /**
+     * Returns an AccessToken as a string
+     *
+     * @return String representing the retrieved AccessToken
+     */
+    String getAccessToken() {
+        this.tokenCredential.getToken(this.context).doOnError(exception -> exception.printStackTrace())
+                .subscribe(token -> {
+            this.accessToken = token;
```

Comment:
`private AccessToken accessToken;` (in line 21) should be scoped within `String getAccessToken()` method since it's only used in the method. This is to avoid unexpected null exceptions in the future.

Reply:
accessToken can only be referenced by the lambda function if it is scoped as a class variable, I can try and find a different work around to scope it down, but whenever I scope down accessToken the lambda function gives an error saying it must be set as a final variable.

Reply:
/**
     * Returns an AccessToken as a string
     *
     * @return String representing the retrieved AccessToken
     */
    String getAccessToken() {
        final AccessToken[] accessToken = new AccessToken[1];
        this.tokenCredential.getToken(this.context).doOnError(exception -> exception.printStackTrace())
                .subscribe(token -> {
            accessToken[0] = token;
        });
        return accessToken[0].getToken();
    }

-This works to scope it down, let me know what you think

Reply:
I see, a neat way would be to make `getAccessToken` return a `Mono` and subscribe to it in `authenticateRequest`, but that's an overkill for this.

The array route seem to be the only feasible route here to avoid a null exception. You can use a String array like this:
``` java
String getAccessToken() {
        String[] accessToken  = new String[0];
        this.tokenCredential.getToken(this.context).doOnError(exception -> exception.printStackTrace())
                .subscribe(token -> {
                    accessToken [0] = token.getToken();
        });
        return accessToken[0];
    }
```